video/envelopes.h

```diff
@@ -0,0 +1,79 @@
+//
+// Title:	        Audio Envelope support
+// Author:        	Steve Sims
+// Created:       	06/08/2023
+// Last Updated:	06/08/2023
+
+class VolumeEnvelope {
+	public:
+		virtual byte getVolume(byte baseVolume, word elapsed, long duration);
+		virtual bool isReleasing(word elapsed, long duration);
+		virtual bool isFinished(word elapsed, long duration);
+};
+
+class ADSRVolumeEnvelope : public VolumeEnvelope {
+	public:
+		ADSRVolumeEnvelope(word attack, word decay, byte sustain, word release);
+		byte getVolume(byte baseVolume, word elapsed, long duration);
+		bool isReleasing(word elapsed, long duration);
+		bool isFinished(word elapsed, long duration);
+	private:
+		word _attack;
+		word _decay;
+		byte _sustain;
+		word _release;
+};
+
+ADSRVolumeEnvelope::ADSRVolumeEnvelope(word attack, word decay, byte sustain, word release) {
+	// attack, decay, release are time values in milliseconds
+	// sustain is 0-255, centered on 127, and is the relative sustain level
+	this->_attack = attack;
+	this->_decay = decay;
+	this->_sustain = sustain;
+	this->_release = release;
+	debug_log("audio_driver: ADSRVolumeEnvelope: attack=%d, decay=%d, sustain=%d, release=%d\n\r", this->_attack, this->_decay, this->_sustain, this->_release);
+}
+
+byte ADSRVolumeEnvelope::getVolume(byte baseVolume, word elapsed, long duration) {
```

Comment:
the `getVolume` method of a volume envelope is used inside the loop to update the current volume level of a note.

when there is no envelope applied to a channel, a note of duration zero will be completely silent, as there is no time for a sound to be made.

when an envelope has been set, the attack and decay phases will always make a sound, but the release phase can be interrupted.

this means that the minimum duration of a note is the total of the attack+decay times, so attempting to play a zero duration note will sound.

to interrupt a note during the attack or decay phases you should clear the envelope.